pkg/querysharding/analyzer.go

```diff
@@ -0,0 +1,106 @@
+package querysharding
+
+import (
+	"fmt"
+
+	"github.com/prometheus/prometheus/promql/parser"
+)
+
+// QueryAnalyzer is an analyzer which determines
+// whether a PromQL Query is shardable and using which labels.
+type QueryAnalyzer struct{}
+
+var nonShardableFuncs = []string{
+	"label_join",
+	"label_replace",
+}
+
+// NewQueryAnalyzer creates a new QueryAnalyzer
+func NewQueryAnalyzer() *QueryAnalyzer {
+	return &QueryAnalyzer{}
+}
+
+// Analyze analyzes a query and returns a QueryAnalysis.
+
+// Analyze uses the following algorithm:
+// * if a query has subqueries, such as label_join or label_replace,
```

Comment:
I guess you want to mention 2 cases here: if a query has subqueries or a query contains non shardable functions then it is not shardable

Reply:
Yes good idea, I amended the comment.